### PR TITLE
Break GroupedHashAggregateStream spill batch into smaller chunks

### LINKED ISSUE #8003: GroupedHashAggregateStream should create smaller spill batches
### Is your feature request related to a problem or challenge?

At the moment GroupedHashAggregateStream will spill state as a single batch, which is not optimal when merging as it loads whole file in memory as a single batch. 

### Describe the solution you'd like

I'd like to spit spill batch into smaller chunks with default batch size same as default batch size set in configuration property.

### Describe alternatives you've considered

I have considered setting batch size to a fixed size or read from configuration property, but at the moment I did not do it as it would be bigger change. 

### Additional context

Relates to #7858

### PR DESCRIPTION
**Pull request by milenkovicm**
...  to decrease memory required for merging them later.

## Which issue does this PR close?

Closes #8003.

## Are there any user-facing changes?

No

### CODE REVIEW

File: datafusion/physical-plan/src/aggregates/row_hash.rs [hunk 1]
```diff
@@ -673,7 +673,18 @@ impl GroupedHashAggregateStream {
         let spillfile = self.runtime.disk_manager.create_tmp_file("HashAggSpill")?;
         let mut writer = IPCWriter::new(spillfile.path(), &emit.schema())?;
         // TODO: slice large `sorted` and write to multiple files in parallel
-        writer.write(&sorted)?;
+        let mut offset = 0;
+        let total_rows = sorted.num_rows();
+
+        while offset < total_rows {
+            // TODO: we could consider smaller batch size as there may be hundreds of batches
```
Comment (alamb) [inline datafusion/physical-plan/src/aggregates/row_hash.rs:680]: this comment seems backwards, shouldn't it be 'consider *larger* batchsize` ?

But in any event, hundreds of batches seems reasonable if they are all 8k rows a piece

cc @kazuyukitanimura

**Review by alamb — APPROVED** [review]
Thank you @milenkovicm

File: datafusion/physical-plan/src/aggregates/row_hash.rs (hunk 1 above)
Comment (milenkovicm) [reply datafusion/physical-plan/src/aggregates/row_hash.rs:680]: There is a typo "...batch size as there may be hundreds of batches" -> "batch size as there may be hundreds of FILES"

What I meant if we have more files spilled, it makes sense, in my head at least, to have smaller batch size as would decrease memory pressure.

Comment (alamb) [reply datafusion/physical-plan/src/aggregates/row_hash.rs:680]: I think I was confused that this says "TODO smaller batch size" as the PR already makes the batch size smaller (caps it at the configured batch size). Given the rest of the engine processes things in the same sized batches, I don't see any reason to use a different batch size for the spill files 🤔

Comment (milenkovicm) [reply datafusion/physical-plan/src/aggregates/row_hash.rs:680]: I have no objection @alamb, will remove the comment

File: datafusion/physical-plan/src/aggregates/mod.rs [hunk 2]
```diff
@@ -2155,7 +2155,7 @@ mod tests {
         spill: bool,
     ) -> Result<()> {
         let task_ctx = if spill {
-            new_spill_ctx(2, 2812)
+            new_spill_ctx(2, 2886)
```
Comment (kazuyukitanimura) [inline datafusion/physical-plan/src/aggregates/mod.rs:2158]: Is this related?

File: datafusion/physical-plan/src/aggregates/row_hash.rs [hunk 3]
```diff
@@ -673,7 +673,16 @@ impl GroupedHashAggregateStream {
...
+        let mut offset = 0;
+        let total_rows = sorted.num_rows();
+
+        while offset < total_rows {
+            let length = std::cmp::min(total_rows - offset, self.batch_size);
+            let batch = sorted.slice(offset, length);
+            offset += batch.num_rows();
+            writer.write(&batch)?;
+        }
+
```
Comment (kazuyukitanimura) [inline datafusion/physical-plan/src/aggregates/row_hash.rs:685]: Is it possible to write in parallel? So that this becomes less blocking.

Additional improvement would be that chunking before sorting. I remember the discussion was `sorted` keeps a copy in memory. We can slice `emit` before `sort_batch` and sort right before writing.

**Review by kazuyukitanimura — COMMENTED** [review]
Thank you @milenkovicm

**Review by Dandandan — APPROVED** [review]

File: datafusion/physical-plan/src/aggregates/mod.rs (hunk 2 above)
Comment (milenkovicm) [reply datafusion/physical-plan/src/aggregates/mod.rs:2158]: It is, with original value it will panic with `ResourcesExhausted` in the merge phase, so I have increase it. Did not do in depth analysis of the problem

File: datafusion/physical-plan/src/aggregates/row_hash.rs (hunk 3 above)
Comment (milenkovicm) [reply datafusion/physical-plan/src/aggregates/row_hash.rs:685]: I don't see big benefit of writing in parallel, but we can give it a try at the later date.
Will try to implement some of the alternatives we discuss, at the moment addressing low hanging fruits.

Comment (kazuyukitanimura) [reply datafusion/physical-plan/src/aggregates/row_hash.rs:685]: Got it.
Would you mind explaining how reading smaller files reduces memory usage? I thought we are streaming when reading back for merging. Also, when merge-sort is going on, we need to open all files anyway? Just making sure the original issue statement in #8003 

Additionally, I think we have to create a new writer otherwise, we are keep appending to the same temp file? So we end up having the same file size?

Comment (milenkovicm) [reply datafusion/physical-plan/src/aggregates/row_hash.rs:685]: Sorry for the confusion, not smaller files, smaller batches. We still keep one file per spill method invocation

Previous implementation would just have one batch per file, this change would introduce more than one batch per file. So file size should be similar size to previous implantation, probably even a bit bigger, but that same file will have more than one batch in it.
 
Streaming merge will open same number of files like before, but it will load smaller batches into memory (if we have one batch per file it would mean whole file would be loaded).

Comment (kazuyukitanimura) [reply datafusion/physical-plan/src/aggregates/row_hash.rs:685]: Thank you for explaining!

**Review by kazuyukitanimura — APPROVED** [review]

**Comment by alamb** [comment]
Thanks everyone
